Add OIDC Auth0 extended tutorial (#35931)
quarkusio/quarkus: docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc

[id="security-oidc-auth0-tutorial"]

> = Protect Quarkus web application by using Auth0 OpenID Connect provider

Reviewer: Missing articles in title.
```suggestion
= Protect a Quarkus web application by using the Auth0 OpenID Connect provider
```

Author: @michelle-purcell Hi Michelle, this suggestion won't work, this is actually referring to an OIDC provider named `Auth0`

include::_attributes.adoc[]
:diataxis-type: tutorial
:categories: security,web

> xref:security-architecture.adoc[Quarkus Security] provides comprehensive OpenId Connect (OIDC) and OAuth2 support with its `quarkus-oidc` extension, supporting both xref:security-oidc-code-flow-authentication.adoc[Authorization code flow] and xref:security-oidc-bearer-token-authentication.adoc[Bearer token] authentication mechanisms.

Reviewer: ```suggestion
xref:security-architecture.adoc[Quarkus Security] provides comprehensive OpenID Connect (OIDC) and OAuth2 support  through the `quarkus-oidc` extension, which supports both xref:security-oidc-code-flow-authentication.adoc[Authorization Code Flow] and xref:security-oidc-bearer-token-authentication.adoc[Bearer token] authentication mechanisms.
```

Author: I'm not sure this one works. I'd like to start with 1 sentence overview of what Quarkus OIDC can do

Well-known OIDC providers such as https://www.keycloak.org/documentation[Keycloak], https://developer.okta.com/[Okta], https://auth0.com/docs/[Auth0] and many xref:security-openid-connect-providers.adoc[social providers] can be configured with ease.

> In this tutorial you will learn how https://auth0.com/docs/[Auth0] can be used to secure your Quarkus endpoints.

Reviewer: Avoid using self-referential statements where possible:
"In this... "

Also, I think line 11 would make a great opening statement, especially on the landing page tile for skimmers:

```suggestion
Learn how to use the Quarkus Security OpenID Connect extension together with https://auth0.com/docs/[Auth0] authorization to secure your API endpoints.
```

Author: This line is a more specific statement, we start with saying Quarkus can support all sort of providers and then say that here we'd look at one of them in particular.

But I can try to use your better worded sentence indeed, with a tiny update:

```Learn how to use the Quarkus OpenID Connect extension (`quarkus-oidc`) together with https://auth0.com/docs/[Auth0] OIDC provider to protect your API endpoints.`

Let me try and see if you will like it

== Troubleshooting

The steps described in this tutorial should work exactly as the tutorial describes. You may have to clear the browser cookies when accessing the updated Quarkus endpoint if you have already completed the authentication. You might need to restart the Quarkus application manually in devmode but it is not expected. Please get in touch with the Quarkus team if you need help completing this tutorial.

== Conclusion

In this tutorial we have looked at how Quarkus endpoints can be secured with Quarkus OIDC (`quarkus-oidc`) extension and `Auth0`, using both Authorization code and Bearer token authentication flows, with both flows supported by the same endpoint code. We have added support for the custom `Auth0` logout flow and enabled Role Based Access Control with a custom `Auth0` namespace qualified claim without writing a single line of code. We used Quarkus devmode to update the code and configuration without restarting the endpoint, and OIDC DevUI to visualize and test `Auth0` tokens.

== References
* xref:security-overview.adoc[Quarkus Security overview]

> * xref:security-oidc-code-flow-authentication.adoc[OIDC code flow mechanism for protecting web applications]

Reviewer: Would it add value to link to the well-known OIDC providers topic, like so:

```suggestion
* xref:security-oidc-code-flow-authentication.adoc[OIDC code flow mechanism for protecting web applications]
* xref:security-openid-connect-providers.adoc[Configuring well-known OpenID Connect providers]
```

[id="security-oidc-auth0-tutorial"]
= Protect Quarkus web application by using `Auth0` OpenID Connect provider
include::_attributes.adoc[]
:diataxis-type: tutorial
:categories: security,web

xref:security-architecture.adoc[Quarkus Security] provides comprehensive OpenId Connect (OIDC) and OAuth2 support with its `quarkus-oidc` extension, supporting both xref:security-oidc-code-flow-authentication.adoc[Authorization code flow] and xref:security-oidc-bearer-token-authentication.adoc[Bearer token] authentication mechanisms.

> Well-known OIDC providers such as https://www.keycloak.org/documentation[Keycloak], https://developer.okta.com/[Okta], https://auth0.com/docs/[Auth0] and many xref:security-openid-connect-providers.adoc[social providers] can be configured with ease.

Reviewer: ```suggestion
With Quarkus, you can easily configure xref:security-openid-connect-providers.adoc[well-known social OIDC and OAuth2 providers], including link:https://www.keycloak.org/documentation[Keycloak], link:https://developer.okta.com/[Okta], and link:https://auth0.com/docs/[Auth0].
```

Author: Hi Michelle, would you be OK if I tried:
```
With Quarkus, you can easily configure OIDC providers such as link:https://www.keycloak.org/documentation[Keycloak], link:https://developer.okta.com/[Okta], and link:https://auth0.com/docs/[Auth0], and xref:security-openid-connect-providers.adoc[well-known social OIDC and OAuth2 providers]
```
?

See, `Keycloak`, `Okta` and `Auth0`, these are not qualified as `social` providers, they are `enterprise` grade (though I would not use `enterprise` here), one thing is certain is that when you see social login options, they'd never offer Keycloak or Okta or Autho :-), while xref:security-openid-connect-providers.adoc (somewhat misnamed file name) lists `social` providers only.

Reviewer: @sberyozkin -  Your suggestion is perfect +1. I see what you mean. Thanks for explaining. Apologies for the churn.

Author: @michelle-purcell Np at all, thanks for trying to improve the text

Learn how to use the Quarkus OpenID Connect extension (`quarkus-oidc`) together with https://auth0.com/docs/[Auth0] OIDC provider to protect your API endpoints.

== Create Auth0 application

> Start with creating an `Auth0` application in the `Auth0` dashboard, for example, lets create a `QuarkusAuth0` web application:

Reviewer: ```suggestion
Start with creating an `Auth0` application in the `Auth0` dashboard, for example, let's create a `QuarkusAuth0` web application:
```
But if following the Quarkus template numbered steps approach:

```suggestion
. Go to the `Auth0` dashboard and create a regular web application. For example, create an `Auth0` application called `QuarkusAuth0`.
```

@sberyozkin - Should I pause suggesting edits to convert the actual steps into numbered sub-steps in this review iteration?   Thanks.

Author: @michelle-purcell I'm applying numbering scheme suggestions, but I'm not sure I'm following where to put `. ` next :-), please propose a few suggestions :-)

> image::auth0-create-application.png[Create Auth0 application]

Reviewer: @sberyozkin - Just some general comments about the screencaps instructional content.

- Screencaps are high maintenance. We typically try to avoid them for both accessibility and localization reasons.
- If the UIs in those screencaps get updated regularly and the docs don't get updated, we risk losing the trust of your reader or creating mismatch confusion.
- I would suggest you use them as judiciously as possible and ensure alt text is unique and descriptive to support visually impaired users.

Reviewer: @sberyozkin - In saying that, _personally as a user/reader, I love them!_

Author: Hi @michelle-purcell I'm not exactly sure I understand, when you say `screencaps`, do you mean the images ? The chance of the text in them changing is very low, if it does it won't change a lot, I've been hoping to  make sure the new Auth0 users who may not be well familiar with the Auth0 setup steps, just follow this tutorial.
We also hope this tutorial be reviewed externally, once published. Ideally, it will be very easy for the reviewers to follow.

Does it clarify why I've added (more than usual) number of images ?

Author: I guess, since you liked it yourself, we can resolve this one :-) ?

Author: @michelle-purcell Let me resolve just to optimize the review space but please re-open if I've missed something

Reviewer: @sberyozkin - Thanks for explaining the rationale here. 

Yes, I mean the screen capture images of the Auth0 UI and other UIs, especially the UI that we don't own. 

If you think they are definitely required and the maintenance risks are low then keep them. 

Important: The screen captures of the UI must be supplemental only for the novice user. So the key question is:
If you removed any of the UI images, would the user have enough information with the text alone to follow and complete the required actions? 

If the answer is no, then we need another iteration to enhance the textual instructions for accessibility support.

For every image in the guide, we also need to ensure that every image has an Alt text attribute that clearly describes the picture for someone who cannot see it. I can do the ALT text review of images after this initial iteration when the step text edits/comments/suggestions are closed out.
Make sense? Thanks.

> `Auth0` will create it and generate the client id and secret and show its HTTPS-based domain:

Reviewer: ```suggestion
.Result
Your`Auth0` application gets created with a client ID, secret, and HTTPS-based domain.
```

image::auth0-created-application.png[Created Auth0 application]

> You will use these 3 properties to configure Quarkus at the next step.

Reviewer: ```suggestion
```

> Next, add users to this application by selecting the `Users` option in the application's dashboard, for example:

Author: @michelle-purcell The step is indeed critical and good point, I suppose I'm hand-holding the reader :-), but I really would like to avoid any problems at all for not too experienced Auth0 users. But indeed, your suggestion is simpler and it is really obvious what to do after reading it :-). Let me apply

image::auth0-add-user.png[Add Auth0 application users]

> Now that the initial `Auth0` application setup has been completed, you are ready to start creating and configuring a Quarkus endpoint. We will be updating the `Auth0` application configuration along the way as necessary.

Reviewer: ```suggestion
Now that you have successfully created and configured your `Auth0` application, you are ready to start creating and configuring a Quarkus endpoint. 
In the steps that follow, you will continue to configure and update your `Auth0` application as needed.
```

> == Create Quarkus application

Reviewer: ```suggestion
== Create a Quarkus application
```

> Use the following Maven command to create a Quarkus REST application which can be secured with its OIDC extension.

Reviewer: Specificity: Should REST be RESTEasy Reactive?

```suggestion
Use the following Maven command to create a Quarkus RESTEasy Reactive application that can be secured with the  Quarkus OIDC extension.
```

Author: Sounds good

:create-app-artifact-id: quarkus-auth0
:create-app-extensions: resteasy-reactive,oidc
include::{includes}/devtools/create-app.adoc[]

> Create the application workspace and import it into your favourite IDE, note a `GreetingResource` JAX-RS resource that will need to be secured:

Reviewer: US spelling.
- Consider rephrasing this sentence. I'm not quite sure what the note is about here.
- Consider using the admonition markup for notes.

```suggestion
Create the application workspace and import it into your favorite IDE.
Note: <rephrase the`GreetingResource` JAX-RS resource sentence>
```

Author: @michelle-purcell I did not mean it to be a `Note`, just wanted to let users know that `GreetingResource` was generated,  which I guess does not warrant a Note-level highlight, can we just replace `note` with some other verb/expression ?

Reviewer: @sberyozkin - I see :+1:  In that case, something like...

```suggestion
Create the application workspace and import it into your favorite IDE.
A `GreetingResource` JAX-RS resource gets generated, which you will need to secure.
```

Not sure if my suggestion here ^^ is technically correct. These questions might help to enhance it further:

- Which step in this procedure section actually generates a `GreetingResource` ?
- When do you need to secure it? 
- Will the reader know how to secure it? (or is it covered in the code sample that follows?)

Thanks

Author: @michelle-purcell Sounds good, yeah, then we'll work in the tutorial to get it secured

[source,java]
----
package org.acme;

import jakarta.inject.Inject;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;

@Path("/hello")
public class GreetingResource {

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        return "Hello from RestEasy Reactive";
    }
}
----

> Now request that only authenticated users can access the `hello()` method and update it to return the current user's name instead:

Reviewer: If this doesn't break the technical meaning...

```suggestion
Update the Quarkus application code so that only authenticated users can access the `hello()` method and update it to return the current username instead, as shown in the following example:
```

[source,java]
----
package org.acme;

import org.eclipse.microprofile.jwt.JsonWebToken;

import io.quarkus.oidc.IdToken;
import io.quarkus.security.Authenticated;
import jakarta.inject.Inject;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;

@Path("/hello")
public class GreetingResource {

    @Inject
    @IdToken                                        <1>
    JsonWebToken idToken;
	
    @GET
    @Authenticated                                  <2>
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        return "Hello, " + idToken.getName();
    }
}
----

> <1> Injected `JsonWebToken` (JWT) token has an `@IdToken` qualifier. `IdToken` represents a user authentication in the OIDC Authorization code flow.

Reviewer: ```suggestion
<1> The injected `JsonWebToken` (JWT) token has an `@IdToken` qualifier.
`IdToken` represents the user authentication part of the OIDC authorization code flow mechanism.
```

<2> `io.quarkus.security.Authenticated` annotation is added to the `hello()` method requiring that only authenticated users can access it.

> Note that the access token acquited during the authorization code flow, alongside the ID token, is not used directly by the endpoint but is only used to access downstream services on behalf of the currently authenticated user. We will talk about using the access tokens later in this tutorial.

Reviewer: ```suggestion
[NOTE]
====
The access token acquired during the authorization code flow, alongside the ID token, is not used directly by the endpoint but is only used to access downstream services on behalf of the currently authenticated user. 
More to come on the topic of "access tokens", later in this tutorial.
```

:sectnums:
:sectnumlevels: 3

== Create an Auth0 application

. Go to the `Auth0` dashboard and create a regular web application. For example, create an `Auth0` application called `QuarkusAuth0`.

image::auth0-create-application.png[Create Auth0 application]

.Result
Your`Auth0` application gets created with a client ID, secret, and HTTPS-based domain.
Make a note of these properties because you will need them to complete the Quarkus configuration in the next step.

image::auth0-created-application.png[Created Auth0 application]

Next, while still in the `Auth0` dashboard, add some users to your application.

image::auth0-add-user.png[Add Auth0 application users]

Now that you have successfully created and configured your `Auth0` application, you are ready to start creating and configuring a Quarkus endpoint. 
In the steps that follow, you will continue to configure and update your `Auth0` application as needed.

== Create a Quarkus application

> Use the following Maven command to create a Quarkus RESTEasy Reactive application that can be secured with the  Quarkus OIDC extension.

Reviewer: Remove extra space, which I introduced :see_no_evil: 

```suggestion
Use the following Maven command to create a Quarkus RESTEasy Reactive application that can be secured with the Quarkus OIDC extension.
```

    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        return "Hello, " + idToken.getName();
    }
}
----
<1> The injected `JsonWebToken` (JWT) token has an `@IdToken` qualifier.
`IdToken` represents the user authentication part of the OIDC authorization code flow mechanism.
<2> The `io.quarkus.security.Authenticated` annotation is added to the `hello()` method, which means that only authenticated users can access it.

[NOTE]
====
The access token acquired during the authorization code flow, alongside the ID token, is not used directly by the endpoint but is only used to access downstream services on behalf of the currently authenticated user. 
More to come on the topic of "access tokens", later in this tutorial.

Configure OIDC in `application.properties` with the properties from the created `Auth0` application:

[source,configuration]
----
# Make sure the application domain is prefixed with 'https://'
quarkus.oidc.auth-server-url=https://dev-3ve0cgn7.us.auth0.com
quarkus.oidc.application-type=web-app
quarkus.oidc.client-id=sKQu1dXjHB6r0sra0Y1YCqBZKWXqCkly
quarkus.oidc.credentials.secret=${client-secret}
----

> Here, you have configured Quarkus to use the created `Auth0` application's domain, client id and secret. `quarkus.oidc.application-type=web-app` is an indication to Quarkus that the OIDC Authorization code flow must be used (we'll consider the alternatives later in this tutorial).

Reviewer: ```suggestion
In completing this step, you have just configured Quarkus to use the domain, client ID, and secret of the `Auth0` application that you created earlier. 
Setting the property `quarkus.oidc.application-type=web-app` instructs Quarkus to use the OIDC authorization code flow but there are also other methods, which are discussed later on in the tutorial.
```

Author: @michelle-purcell `In completing this step` definitely reads better than `Here` :-)

> Note that you will use Quarkus `devmode` to test the endpoint so the endpoint address will be `http://localhost:8080/hello`. You need to register this address as an allowed callback URL in your `Auth0` application:

Reviewer: ```suggestion
To test the endpoint by using Quarkus dev mode, you must complete the following steps:
* Set the endpoint address to \http://localhost:8080/hello. 
* Register \http://localhost:8080/hello as an allowed callback URL in your `Auth0` application.
```

In AsciiDoc, you can use a backslash (\) char to escape URLs so that they don't become active links. The \ isn't rendered in the output.

Author: ```suggestion
The endpoint address will be  \http://localhost:8080/hello and you must register this address as an allowed callback URL in your `Auth0` application.
```

@michelle-purcell I've tried to simplify - I put quite a bit irrelevant content earlier overloading a simple instruction (ex. the same address will be in the prod mode)

Reviewer: @sberyozkin - I see. How about this minor tweak to your suggested edits?

```suggestion
By default, the endpoint address for the application is set to  \http://localhost:8080/hello so you must register this address as an allowed callback URL in your `Auth0` application.
```

image::auth0-allowed-callback.png[Auth0 allowed callback URL]

> This means that you will access the Quarkus `http://localhost:8080/hello` endpoint from the browser and `Auth0` will redirect you back to the same address after the authentication has been completed.

Reviewer: ```suggestion
Now when you access the Quarkus \http://localhost:8080/hello endpoint from the browser, `Auth0` redirects you back to the same address after the authentication is completed.
```

[id="security-oidc-auth0-tutorial"]

> = Protect Quarkus web application by using `Auth0` OpenID Connect provider

Reviewer: Missing article. 

```suggestion
= Protect Quarkus web application by using an `Auth0` OpenID Connect provider
```

include::_attributes.adoc[]
:diataxis-type: tutorial
:categories: security,web

xref:security-architecture.adoc[Quarkus Security] provides comprehensive OpenId Connect (OIDC) and OAuth2 support with its `quarkus-oidc` extension, supporting both xref:security-oidc-code-flow-authentication.adoc[Authorization code flow] and xref:security-oidc-bearer-token-authentication.adoc[Bearer token] authentication mechanisms.

With Quarkus, you can easily configure OIDC providers such as link:https://www.keycloak.org/documentation[Keycloak], link:https://developer.okta.com/[Okta], link:https://auth0.com/docs/[Auth0], and other xref:security-openid-connect-providers.adoc[well-known social OIDC and OAuth2 providers].

Learn how to use the Quarkus OpenID Connect extension (`quarkus-oidc`) together with the https://auth0.com/docs/[Auth0] OIDC provider to protect your API endpoints.

== Prerequisites

> Reviewing https://auth0.com/docs/[Auth0 documentation] is recommended.

Reviewer: ```suggestion
Before you begin, review the link:https://auth0.com/docs/[Auth0] documentation.
```

> Reviewing https://auth0.com/docs/[Auth0 documentation] is recommended.

Reviewer: Merge this into 16 as per suggestion below.
```suggestion
```

> Reviewing Quarkus OIDC documents about xref:security-oidc-code-flow-authentication.adoc[Authorization code flow] and xref:security-oidc-bearer-token-authentication.adoc[Bearer token] authentication mechanisms will make it easier for you to complete this tutorial.

Reviewer: How about merging >15 and >16 into one list of recommended reading?

```suggestion
To make it easier for you to complete this tutorial, review the following documentation before you begin:

* link:https://auth0.com/docs/[Auth0 docs site]
* xref:security-oidc-code-flow-authentication.adoc[Quarkus OpenID Connect authorization code flow mechanism for protecting web applications]
* xref:security-oidc-bearer-token-authentication.adoc[Quarkus OpenID Connect (OIDC) Bearer token authentication]
```

:sectnums:
:sectnumlevels: 3

== Create an Auth0 application

> . Go to the `Auth0` dashboard and create a regular web application. For example, create an `Auth0` application called `QuarkusAuth0`.

Reviewer: Looks odd having only 1 numbered step in this procedure.

Either make all actions part of an ordered list with numbered steps in sequence or remove this lone numbered step.

The tutorial template encourages the numbering of all action steps in a procedure, however this approach isn't widely adopted by contributors from what I can see.

Author: @michelle-purcell Right, I'm not I was understanding it would add numbers like `1.`, I'd just remove it

> . Go to the `Auth0` dashboard and create a regular web application. For example, create an `Auth0` application called `QuarkusAuth0`.

Reviewer: ```suggestion
. Go to the `Auth0` dashboard and create a regular web application. 
For example, create an `Auth0` application called `QuarkusAuth0`.
```

image::auth0-create-application.png[Create Auth0 application]

.Result

> Your`Auth0` application gets created with a client ID, secret, and HTTPS-based domain.

Reviewer: Does Auth0 need to be inside backticks?
It's a valid proper noun. If Vale rejects it, I can fix that rule to remove the false positive.

```suggestion
Your Auth0 application gets created with a client ID, secret, and HTTPS-based domain.
```

Make a note of these properties because you will need them to complete the Quarkus configuration in the next step.

image::auth0-created-application.png[Created Auth0 application]

> Next, while still in the `Auth0` dashboard, add some users to your application.

Reviewer: ```suggestion
Next, while still in the Auth0 dashboard, add some users to your application.
```

    }
}
----

Update the Quarkus application code so that only authenticated users can access the `hello()` method and update it to return the current username instead, as shown in the following example:

[source,java]
----
package org.acme;

import org.eclipse.microprofile.jwt.JsonWebToken;

import io.quarkus.oidc.IdToken;
import io.quarkus.security.Authenticated;
import jakarta.inject.Inject;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;

@Path("/hello")
public class GreetingResource {

    @Inject
    @IdToken                                        <1>
    JsonWebToken idToken;
	
    @GET
    @Authenticated                                  <2>
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        return "Hello, " + idToken.getName();
    }
}
----
<1> The injected `JsonWebToken` (JWT) token has an `@IdToken` qualifier.
`IdToken` represents the user authentication part of the OIDC authorization code flow mechanism.
<2> The `io.quarkus.security.Authenticated` annotation is added to the `hello()` method, which means that only authenticated users can access it.

[NOTE]
====

> The access token acquired during the authorization code flow, alongside the ID token, is not used directly by the endpoint but is only used to access downstream services on behalf of the currently authenticated user. 

Reviewer: Misplaced only:

```suggestion
The access token acquired during the authorization code flow, alongside the ID token, is not used directly by the endpoint but is used only to access downstream services on behalf of the currently authenticated user. 
```

More to come on the topic of "access tokens", later in this tutorial.
====

> Configure OIDC in `application.properties` with the properties from the created `Auth0` application:

Reviewer: As there are two platforms being used in this tutorial, would it help to set the context in places, for example:

```suggestion
Configure OIDC in the Quarkus `application.properties` file  by using the properties from the Auth0 application that you created earlier.
```

[source,configuration]
----
# Make sure the application domain is prefixed with 'https://'
quarkus.oidc.auth-server-url=https://dev-3ve0cgn7.us.auth0.com
quarkus.oidc.application-type=web-app
quarkus.oidc.client-id=sKQu1dXjHB6r0sra0Y1YCqBZKWXqCkly
quarkus.oidc.credentials.secret=${client-secret}
----

> In completing this step, you have just configured Quarkus to use the domain, client ID, and secret of the `Auth0` application that you created earlier. 

Reviewer: Same comment about backticks around Auth0:

```suggestion
In completing this step, you have just configured Quarkus to use the domain, client ID, and secret of yourAuth0 application.
```

> Setting the property `quarkus.oidc.application-type=web-app` instructs Quarkus to use the OIDC authorization code flow but there are also other methods, which are discussed later on in the tutorial.

Reviewer: ```suggestion
Setting the property `quarkus.oidc.application-type=web-app` instructs Quarkus to use the OIDC authorization code flow, but there are also other methods, which are discussed later on in the tutorial.
```

> The endpoint address will be  \http://localhost:8080/hello and you must register this address as an allowed callback URL in your `Auth0` application.

Reviewer: ```suggestion
The endpoint address will be  \http://localhost:8080/hello, which must also be registered as a permitted callback URL in your Auth0 application.
```

image::auth0-allowed-callback.png[Auth0 allowed callback URL]

> Now when you access the Quarkus \http://localhost:8080/hello endpoint from the browser, `Auth0` redirects you back to the same address after the authentication is completed.

Reviewer: ```suggestion
After completing this step. when you access the Quarkus \http://localhost:8080/hello endpoint from a browser, Auth0 redirects you back to the same address after the authentication is completed.
```

[NOTE]
====

> Quarkus also allows to configure the callback path with the `quarkus.oidc.authentication.redirect-path` property. Why do you not have to configure a callback path for this Quarkus application ? By default, Quarkus will use the current request path as the callback path which is why you don't have to configure `quarkus.oidc.authentication.redirect-path=/hello`.

Reviewer: Could we simplify these 2 sentences:

```suggestion
By default, Quarkus automatically uses the current request path as the callback path.
But you can override the default behavior and configure a specific callback path by setting the Quarkus `quarkus.oidc.authentication.redirect-path` property, as outlined in the following example:

`quarkus.oidc.authentication.redirect-path=/hello`
```

quarkus.oidc.client-id=sKQu1dXjHB6r0sra0Y1YCqBZKWXqCkly
quarkus.oidc.credentials.secret=${client-secret}
----

In completing this step, you have just configured Quarkus to use the domain, client ID, and secret of yourAuth0 application.
Setting the property `quarkus.oidc.application-type=web-app` instructs Quarkus to use the OIDC authorization code flow, but there are also other methods, which are discussed later on in the tutorial.

The endpoint address will be  \http://localhost:8080/hello, which must also be registered as a permitted callback URL in your Auth0 application.

image::auth0-allowed-callback.png[Auth0 allowed callback URL]

After completing this step, when you access the Quarkus \http://localhost:8080/hello endpoint from a browser, Auth0 redirects you back to the same address after the authentication is completed.

[NOTE]
====
By default, Quarkus automatically uses the current request path as the callback path.
But you can override the default behavior and configure a specific callback path by setting the Quarkus `quarkus.oidc.authentication.redirect-path` property, as outlined in the following example:

`quarkus.oidc.authentication.redirect-path=/hello`

> In production, your application will most likely have a larger URL space, with many more endpoint addresses available. In such cases you will configure Quarkus with a dedicated callback (redirect) path, for example, `quarkus.oidc.authentication.redirect-path=/authenticated-welcome` and register this URL in the provider's dashboard as `Auth0` will not allow an open ended list of redirect URLs. Quarkus will call `/authenticated-welcome` after accepting a redirect from `Auth0` and completing the Authorization code flow and creating the session cookie - and now your application can let authenticated users access other parts of the secured application space, for example, by using JAX-RS API to redirect, without requiring users to reauthenticate.

Author: Thanks @michelle-purcell. reads OK

====

Now you are ready to start testing the endpoint.

> == Test Quarkus endpoint - first try

Reviewer: ```suggestion
== Test the Quarkus endpoint
```

> Start Quarkus in `devmode`:

Reviewer: ```suggestion
Start Quarkus in dev mode:
```

[source,bash]
----
$ mvn quarkus:dev
----

> This is the only time you are expected to start Quarkus manually in devmode while working with this tutorial. All the configuration and code updates in the follow up sections is expected to be recognized by Quarkus without having to restart the application manually.

Reviewer: ```suggestion
[NOTE]
====
This is the only time during this tutorial that you will need to manually start Quarkus in dev mode.
The configuration and code update steps in the remaining sections of this tutorial are automatically observed and processed by Quarkus without you needing to restart the application manually.
====
```

> Open the browser and access `http://localhost:8080/hello`, you will be redirected to `Auth0` where you will login:

Reviewer: ```suggestion
Open the browser and access the following URL:
`http://localhost:8080/hello`
You will be redirected to Auth0 and prompted to  log in:
```

image::auth0-login.png[Auth0 Login]

> and authorize the `QuarkusAuth0` application to access your account:

Reviewer: ```suggestion
Next, give the `QuarkusAuth0` application authorization to access your account:
```

Author: @michelle-purcell It reads a little bit unusual in scope of describing the OIDC code flow, though it is surely a perfect text as is :-), I'll keep the original text if you don't mind - this is a typical combination AFAIK, `user authorizes` or `user grants permissions`

Reviewer: :+ - One thing: fix capitalization of the sentence. How about:

```suggestion
Authorize the `QuarkusAuth0` application to access your account:
```

image::auth0-authorize.png[Auth0 Authorize]

> and finally you will be redirected back to the Quarkus endpoint and get a response, `Hello, auth0|60e5a305e8da5a006aef5471`.

Reviewer: ```suggestion
Finally, you will be redirected back to the Quarkus endpoint with the following response:
`Hello, auth0|60e5a305e8da5a006aef5471`
```

> This is not the current user's name though, so why has this strange sequence been returned ?

Reviewer: ```suggestion
[TIP]
====
Notice that the current username does not get returned. 
To learn more about why this behavior occurs, see the xref:security-openid-connect-dev-services.adoc#dev-ui-all-oidc-providers[Dev UI for all OpenID Connect Providers] section of the "Dev Services and UI for OpenID Connect (OIDC)" guide.
====
```

> This is not the current user's name though, so why has this strange sequence been returned ?

Reviewer: Replace with suggestion in line 185:

```suggestion
```

> Lets use https://quarkus.io/guides/security-openid-connect-dev-services#dev-ui-all-oidc-providers[Quarkus OIDC Dev UI console] to find out.

Reviewer: ```suggestion
Let's use link:https://quarkus.io/guides/security-openid-connect-dev-services#dev-ui-all-oidc-providers[Quarkus OIDC Dev UI console] to find out.
```

> Lets use https://quarkus.io/guides/security-openid-connect-dev-services#dev-ui-all-oidc-providers[Quarkus OIDC Dev UI console] to find out.

Reviewer: See suggestion above to replace this phrase:

```suggestion
```

> == Looking at Auth0 tokens in OIDC DevUI

Reviewer: ```suggestion
== Looking at Auth0 tokens in the OIDC Dev UI
```

> Quarkus provides a great https://quarkus.io/guides/dev-ui-v2[DevUI] experience. In particular, it offers an out of the box support for developing and testing OIDC endpoints with a Keycloak container, https://quarkus.io/guides/security-openid-connect-dev-services#dev-services-for-keycloak[DevService for Keycloak] which is launched when the OIDC provider's address is not already configured with `quarkus.oidc.auth-server-url`.

Author: @michelle-purcell Sure, I was trying to use `xref` but forgot in this case (hopefully only in this case) :-)

> However, Quarkus users can continue using OIDC DevUI when the provider is already configured. Lets check it right now.

Reviewer: ```suggestion
However, Quarkus users can continue using the OIDC Dev UI when the provider is already configured. 
Use the following instructions to check this configuration:
```

> First though, make sure you change your Quarkus application to have a `hybrid` application type, instead of the current `web-app`:

Reviewer: ```suggestion
First, change your Quarkus application type from `web-app` to `hybrid`, as follows:
```

[source,configuration]
----
quarkus.oidc.auth-server-url=https://dev-3ve0cgn7.us.auth0.com
quarkus.oidc.application-type=hybrid <1>
quarkus.oidc.client-id=sKQu1dXjHB6r0sra0Y1YCqBZKWXqCkly
quarkus.oidc.credentials.secret=${client-secret}
----

> <1> Application type is changed to `hybrid`. The reason it was changed to `hybrid` is because OIDC Dev UI currently supports an `SPA` mode only, i.e, it itself, using its own Java Script, authenticates users to the OIDC provider, and uses the access token as a `Bearer` token to access the Quarkus endpoint as a service.

Reviewer: ```suggestion
<1> Application type is changed to `hybrid` because OIDC Dev UI currently supports `SPA` mode only.
For example, using its own Java Script, the application authenticates users to the OIDC provider and uses the access token as a Bearer token to access the Quarkus endpoint as a service.
```

Author: @michelle-purcell It makes it not quite obvious which application uses `JavaScript`, let me tweak it a bit

> Typically, Quarkus has to be configured with `quarkus.oidc.application-type=service` to support `Bearer` token authentication but it also supports a `hybrid` application type which means it can support both the Authorization code flow and Bearer token flows at the same time.

Reviewer: ```suggestion
Typically, Quarkus must be configured with `quarkus.oidc.application-type=service` to support `Bearer` token authentication, but it also supports a `hybrid` application type, which means it can support both the Authorization code flow and Bearer token flows at the same time.
```

> The other thing you need to do is to configure the Auth0 application to allow the callbacks to OIDC Dev UI, its URL format is  `http://localhost:8080/q/dev-ui/io.quarkus.quarkus-oidc/${provider-name}-provider` where in this case the `${provider-name}` is `auth0`:

Reviewer: ```suggestion
You also need to configure the Auth0 application to allow the callbacks to the OIDC Dev UI.
Use the following URL format:

`http://localhost:8080/q/dev-ui/io.quarkus.quarkus-oidc/${provider-name}-provider` 

* Where in this example, the `${provider-name}` is `auth0`
```

image::auth0-allowed-callbacks.png[Auth0 Allowed Callbacks]

> Now you are ready to use OIDC Dev UI with Auth0, open `http://localhost:8080/q/dev/` in the browser and you will find an OpenId Connect card linking to an `Auth0` provider SPA:

Reviewer: ```suggestion

Now you are ready to use OIDC Dev UI with Auth0.

Open `http://localhost:8080/q/dev/` in a browser session. An OpenId Connect card that links to an Auth0 provider SPA displays, as follows:
```

image::auth0-devui.png[Auth0 DevUI]

> Click on the `Auth0 provider` link and press `Login into Single Page Application`:

Reviewer: ```suggestion
Click *Auth0 provider* followed by *Login into Single Page Application*:
```

image::auth0-devui-login-to-spa.png[Auth0 DevUI Login to SPA]

> You will be redirected to `Auth0`, where you will login and be redirected to the OIDC Dev UI Dashboard:

Reviewer: ```suggestion
You will be redirected to `Auth0`, where you will log in and be redirected to the OIDC Dev UI Dashboard:
```

:sectnums:
:sectnumlevels: 3

== Create an Auth0 application

Go to the `Auth0` dashboard and create a regular web application. 
For example, create an `Auth0` application called `QuarkusAuth0`.

image::auth0-create-application.png[Create Auth0 application]

.Result
Your Auth0 application gets created with a client ID, secret, and HTTPS-based domain.
Make a note of these properties because you will need them to complete the Quarkus configuration in the next step.

image::auth0-created-application.png[Created Auth0 application]

Next, while still in the Auth0 dashboard, add some users to your application.

image::auth0-add-user.png[Add Auth0 application users]

Now that you have successfully created and configured your `Auth0` application, you are ready to start creating and configuring a Quarkus endpoint. 
In the steps that follow, you will continue to configure and update your `Auth0` application as needed.

== Create a Quarkus application

Use the following Maven command to create a Quarkus RESTEasy Reactive application that can be secured with the Quarkus OIDC extension.

:create-app-artifact-id: quarkus-auth0
:create-app-extensions: resteasy-reactive,oidc

> include::{includes}/devtools/create-app.adoc[]

Author: This sequence creates a project without the generated code, so somehow it needs to be tweaked to have the starter code generated

We can fix it by telling Quarkus OIDC which claim must be checked to enforce RBAC:

[source,configuration]
----
quarkus.oidc.auth-server-url=https://dev-3ve0cgn7.us.auth0.com
quarkus.oidc.application-type=hybrid
quarkus.oidc.authentication.scopes=profile
quarkus.oidc.client-id=sKQu1dXjHB6r0sra0Y1YCqBZKWXqCkly
quarkus.oidc.credentials.secret=${client-secret}

quarkus.oidc.roles.role-claim-path="https://quarkus-security.com/roles" <1>

# Logout
quarkus.oidc.end-session-path=v2/logout
quarkus.oidc.logout.post-logout-uri-param=returnTo
quarkus.oidc.logout.extra-params.client_id=${quarkus.oidc.client-id}
quarkus.oidc.logout.path=/logout
quarkus.oidc.logout.post-logout-path=/hello/post-logout
quarkus.http.auth.permission.authenticated.paths=/logout
quarkus.http.auth.permission.authenticated.policy=authenticated
----
<1> Point to the custom roles claim. The path to the roles claim is in double quotes because the claim is namespace qualified.

Now, clear the browser cookie cache just in case, open `http://localhost:8080/hello` again, authenticate to `Auth0` and get an expected user name.

> == Access Quarkus with Auth0 access tokens

Author: Note somewhere that if the access token is also in the JWT format then it can be verified directly
